### ISSUE TITLE
PIT Threading.Sleep() causes stack overflow

### ISSUE THREAD

**Issue by Arawn-Davies** [CosmosOS/Cosmos#1856]
#### Area of Cosmos - What area of Cosmos are we dealing with?
Kernel - CPU SysThreading.Sleep()

#### Expected Behaviour - What do you think that should happen?
Invoking System.Threading.Thread.Sleep(3000); should make the PIT sleep for the specified param (3000)

#### Actual Behaviour - What unexpectedly happens?
Stackoverflow, at the address 
```
020d40bc l       .text	00000000 METHOD_4350739.StackOverflowCheck_GetAddress
```

#### Reproduction - How did you get this error to appear?
Any use of Thread.Sleep()

#### Version - Were you using the User Kit or Dev Kit? And what User Kit version or Dev Kit commit (Cosmos, IL2CPU, X#)?
Devkit latest master

**Comment by Arawn-Davies** [comment]
Of my Kernel.Map file, the stackoverflow occurs at line 9384:
![image](https://user-images.githubusercontent.com/19764552/123338008-04cb7d00-d540-11eb-8b89-e12f29067acd.png)

and the debugger reports back this error:
![image](https://user-images.githubusercontent.com/19764552/123338053-1c0a6a80-d540-11eb-927e-e66152035374.png)

**Comment by Arawn-Davies** [comment]
The PIT.cs class hasn't been modified for 4 years and it has been working in previous userkits, so the issue lies in a separate class elsewhere...

Can confirm that the bug occurs regardless of VM, but a different error occurs depending on whether Debugging is enabled
Stack overflow without remote debugging:
![Uploading image.png…]()

**Comment by AcaiBerii** [comment]
This might be similar to #1836.

**Comment by quajak** [comment]
Does using PIT.Sleep directly work? Could you post the emitted asm for SystemThreadingSleep?

**Comment by Arawn-Davies** [comment]
> This might be similar to #1836.

It is, but I opened a new issue in case there's differences in the Plugged method and using PIT directly

**Comment by Arawn-Davies** [comment]
> Does using PIT.Sleep directly work? Could you post the emitted asm for SystemThreadingSleep?

PIT.Wait works fine, but I'm not sure how accurate it is, I don't understand how just plugging it is causing the SOF? Even just changing Global.PIT.Wait() for System.Threading.Thread.Sleep() with the same params seems to be the only difference.

Here's the emitted asm anyway
```assembly
SystemVoidSystemThreadingThreadSleepSystemInt32:

		GUID_4323774:
			mov EAX, Before_Kernel_Stack
			cmp EAX, ESP
			JL near METHOD_4323774.StackOverflowCheck_End
			cli
			Call METHOD_4323774.StackOverflowCheck_GetAddress

		METHOD_4323774.StackOverflowCheck_GetAddress:
			pop dword EAX
			mov dword [DebugStub_CallerEIP], EAX
			Call DebugStub_SendStackOverflowOccurred
			hlt

		METHOD_4323774.StackOverflowCheck_End:
			push dword EBP
			mov EBP, ESP

		SystemVoidSystemThreadingThreadSleepSystemInt32.IL_0000:
			; Ldarg
			; Arg idx = 0
			; Arg type = System.Int32
			; Arg real size = 4 aligned size = 4
			push dword [EBP + 8]

		SystemVoidSystemThreadingThreadSleepSystemInt32.IL_0001:
			; Call
			Call SystemVoidInternalRuntimeAugmentsRuntimeThreadSleepSystemInt32
			test ECX, 0x2
			JE near SystemVoidSystemThreadingThreadSleepSystemInt32.IL_0006
			; aStackSizeBeforeCall = 4
			; aTotalArgumentSizeOfMethod = 4
			; aReturnSize = 0
			JNE near SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_EXCEPTION

		SystemVoidSystemThreadingThreadSleepSystemInt32.IL_0006:
			; Ret
			Jmp SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_NORMAL
			; End Method: Sleep

		SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_NORMAL:
			; Following code is for debugging. Adjust accordingly!
			mov dword [INTs_LastKnownAddress], SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_NORMAL
			mov ECX, 0x0

		SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_EXCEPTION:
			mov EAX, ESP
			mov EBX, EBP
			cmp EAX, EBX
			JE near SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_EXCEPTION__2
			cli
			Call SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_EXCEPTION.MethodFooterStackCorruptionCheck_Break_on_location

		SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_EXCEPTION.MethodFooterStackCorruptionCheck_Break_on_location:
			pop dword ECX
			push dword EAX
			push dword EBX
			mov dword [DebugStub_CallerEIP], ECX
			Call DebugStub_SendSimpleNumber
			add ESP, 0x4
			Call DebugStub_SendSimpleNumber
			add ESP, 0x4
			Call DebugStub_SendStackCorruptionOccurred
			hlt

		SystemVoidSystemThreadingThreadSleepSystemInt32.END__OF__METHOD_EXCEPTION__2:
			pop dword EBP
			Ret 0x4
```

**Comment by Arawn-Davies** [comment]
![PIT Ticking away...](https://gyazo.com/2c23827144bc49286907f2cad1b18d01.gif)

**Comment by quajak** [comment]
Looks like it's not using the plug. Don't know why though. Maybe add a PlugMethod to the methods directly so that it uses the plug.

**Comment by Arawn-Davies** [comment]
Fixed in #1861
